# Problem Set Examples

This is a *Jupyter* notebook using interactive *Python*. 

We can use this notebook as a calculator. The benefit is that all **our calculations can be documented** so that we can easily find errors or make changes without starting from the beginning. I recommend using *Python* notebooks for all your math. There are thousands of tools available through *Python* and we can go from the simple math of this notebook all the way to powerful data science and AI projects.

We will demonstrate the last three problems from the "**Math Ideas for Biochemistry - Part 1**" document available on the moodle site.

## Problem 8 from Math Ideas

I add 10 mL of 0.1 M HCl to 170 mL of pure water. What is the *p*H of this new solution? 

**Solution**

This is a simple $\rm c_1v_1 = c_2v_2$ problem with the addition of converting your answer to $\rm pH$.

The initial colume, $\rm c_1$, is 10 mL. The final volume, $\rm C_2$, will be 180 mL (170 mL + 10 mL). The initial concentration of HCl is 0.1 M. HCl is a very strong acid and will be 100% dissociated, so we can say that the hydronium ion concentration will be the concentration of HCl.

Examine the *Python* commands below. Observe how they are no different than using a calculator.

In [1]:
import numpy as np    # import additional math tools

v1 = 10 # mL
v1 = v1 / 1000  # convert mL to L
c1 = 0.1 # moles/L

volumeWater = 170 # mL

v2 = volumeWater + v1
v2 = v2 / 1000  # convert mL to L

c2 = v1 * c1 / v2

pH = -np.log10(c2)

print(f"the pH is {pH:0.2f}")

the pH is 2.23


In [2]:
# You dont have to use the fancy print command above.  You can just call the value directly if you want.

pH

2.255272505103306

## Problem 9 from Math Ideas

I have 230 mL of a solution of pure water with enough HCl added to make the *p*H equal to 4.5. I add 10 mL of a 0.15 M NaOH solution. What is the *p*H now? 

**Solution**

We have a volume and concentration for an HCl solution. $\rm [HCl] = [H^+]$ and we can get $\rm [H^+]$ from the $\rm pH$ value.

Obtain the number of moles of HCl present, subtract the number of moles of NaOH added, and then divide by the new volume to get the concentration of HCl remaining. Then convert to $\rm pH$.

In [2]:
pH = 4.5
v1 = 230 # mL
v1 = v1/1000 # convert mL to L
c1 = 10**(-pH) # Units will be moles/L

moles_HCl = c1 * v1

volumeNaOHsolution = 10 # mL
volumeNaOHsolution = volumeNaOHsolution/1000 # convert mL to L
concNaOHsolution = 0.15 # moles/L

molesNaOH = volumeNaOHsolution * concNaOHsolution

molesBase = molesNaOH - moles_HCl

concBase = molesBase / (v1 + volumeNaOHsolution)

pOH = -np.log10(concBase)
pH = 14 - pOH

print(f"the pH is {pH:0.2f}")

the pH is 11.79


## Problem 10 from Math Ideas

I have 230 mL of a solution of 0.2 M acetic acid and have added enough sodium hydroxide solution to set the $\rm pH$ to 4.5. The $\rm pK_a$ of acetic acid is 4.7. 

1. What are the concentrations of acetate ion and acetic acid in this solution? 
2. I add 10 mL of a 0.15 M NaOH solution. What is the $\rm pH$ now? 

**Solution**

Calculate the moles of Acetic acid (the acidic for, HOAc) in the sample and compare it to the moles of NaOH to be added.  We will see that there will be an excess of NaOH. So I subtract [HOAc] from [NaOH] to get the concentration of hydroxide ion in the new solution. I then obtained the $\rm pOH$, and from that, the $\rm pH$


In [3]:
pHsolution = 4.5
volumeSolution = 230 # mL
volumeSolution = volumeSolution/1000 # convert mL to L
concAceticAcid = 0.2 # moles/L
molesAceticAcid = concAceticAcid * volumeSolution

pKaAceticAcid = 4.7

# Use Henderson-Hasselbalch
# pH = pKa + log10(A/HA)
#
# A/HA = 10**(pH - pKa)

A_HA_ratio = 10**(pHsolution - pKaAceticAcid)
print(f"the A/AH ratio is {A_HA_ratio:0.3f}")

moleFractionA = A_HA_ratio/(A_HA_ratio+1)
moleFractionHA = 1/(A_HA_ratio+1)

volumeNaOHsolution = 10 # mL
volumeNaOHsolution = volumeNaOHsolution/1000 # convert mL to L
concNaOHsolution = 0.15 # moles/L

molesNaOH = volumeNaOHsolution * concNaOHsolution
print(f"molesNaOH = {molesNaOH:0.4f}")


molesA = molesAceticAcid * moleFractionA
molesHA = molesAceticAcid * moleFractionHA

print(f"molesA = {molesA:0.3f}")
print(f"molesHA = {molesHA:0.3f}")
print(f"It will take {molesHA/concNaOHsolution*1000:0.1f} mL of NaOH solution to overcome the buffer completely")

molesA = molesA + molesNaOH
molesHA = molesHA - molesNaOH

print(f"After NaOH molesA = {molesA:0.3f}")
print(f"After NaOH molesHA = {molesHA:0.3f}")

A_HA_ratio = molesA/molesHA
print(f"After NaOH the A/AH ratio is = {A_HA_ratio:0.3f}")


pH = pKaAceticAcid + np.log10(A_HA_ratio)
print()
print(f"the pH is {pH:0.2f}")

the A/AH ratio is 0.631
molesNaOH = 0.0015
molesA = 0.018
molesHA = 0.028
It will take 188.0 mL of NaOH solution to overcome the buffer completely
After NaOH molesA = 0.019
After NaOH molesHA = 0.027
After NaOH the A/AH ratio is = 0.723

the pH is 4.56


## Making a Buffer for the Assignment

The math below involves making a buffer for a biochemical experiment. In this example, I will be using HEPES buffer. It is an organic molecule that is compatible with cells and the acidic form has a $\rm pK_a$ value of $7.43$. I will purchase a bottle of crystaline [HEPES (the acidic form)](https://www.sigmaaldrich.com/CA/en/substance/hepes238307365459) and a bottle of crystaline [HEPES sodium salt (the basic form)](https://www.sigmaaldrich.com/CA/en/substance/hepessodiumsalt2602975277393). I want to make a $1.5$ L volume with a total HEPES concentration of $200$ mM and a $\rm pH$ of $7.15$.

We know the $\rm pH$, the $\rm pK_a$ of the acid form of HEPES and the molar mass of both materials (just click on the links in the above paragraph to get more information about both).

We can get the needed ratio of the two forms, the value of $\frac{[A]}{[HA]}$ required if $\rm pH = 7.15$ by using the Hendersson-Haselbalch equation. 

$$pH = pK_a + \log{\frac{[A]}{[HA]}}$$

and we want the total concentration of both forms of HEPES to be $200$ mM so we can say that

$$[A]+[HA]=0.200$$

The code below will serve as a calculator for our math.  Observe how what you do is recorded in the text of the code itself. You can come back to this code and change some numbers to use it again. Feel free to **steal it** and modify it for your own needs.



In [14]:
#######################################################################
#### Set up variables. Change these if I want something different
#######################################################################

pH = 7.15             #### chosen pH
pKa = 7.45            #### pKa value of acid form of buffer. I looked it up on wikipedia
  
MolMass_HA = 238.30   #### g/mole. Molar mass of acidic form. I looked it up in the specification sheet available from the vendor
MolMass_A = 260.29    #### g/mole. Molar mass of sodium salt of HEPES (basic form). I looked it up in the specification sheet available from the vendor
  
Total_Conc = 0.200    #### mole/L.  Desired total concentration.
  
Volume = 1.5          #### L. Volume of buffer solution.

#######################################################################
#### Calculate A/HA ratio
#######################################################################

ratio_A_HA = 10 ** (pH - pKa)            ### Reorganized HH eqn to solve for A/HA

print(f"Ratio of A/HA is {ratio_A_HA:0.3f}")
print()

#######################################################################
#### Calculate A and HA concentrations
#######################################################################

r = ratio_A_HA       #### so that the code below will be easier to read

percent_A =  r / (1 + r)  #### Convert ratio to mole fraction (percent)
percent_HA = 1 / (1 + r)

conc_HA = Total_Conc * percent_HA   #### the fraction of total conc is the conc of each.
conc_A = Total_Conc * percent_A

print(f"Concentration of A is {conc_A:0.3f} moles/L")
print(f"Concentration of HA is {conc_HA:0.3f} moles/L")

print(f"Total concentration of HA + A is {conc_HA+conc_A:0.3f} moles/L")   #### Check. Just in case.
print()

#######################################################################
#### Calculate A and HA amounts
#######################################################################

mass_of_A = conc_A * Volume * MolMass_A       #### moles = conc * volume.  mass = moles * molar_mass
mass_of_HA = conc_HA * Volume * MolMass_HA

print(f"Mass of A in {Volume:0.1f}L is {mass_of_A:0.2f} grams")
print(f"Mass of HA in {Volume:0.1f}L is {mass_of_HA:0.2f} grams")



Ratio of A/HA is 0.501

Concentration of A is 0.067 moles/L
Concentration of HA is 0.133 moles/L
Total concentration of HA + A is 0.200 moles/L

Mass of A in 1.5L is 26.07 grams
Mass of HA in 1.5L is 47.62 grams


### Making the Buffer

We see from the math above that I will need $47.62$ grams of the HEPES acid and $26.07$ grams of its sodium salt dissolved in a final volume of $1.5$ litres. Now I just have to write up a recipe. Take note of how the code above is just calculations steps that are recorded as $Python* commands and can be checked and corrected easily by the reader. Try seeing what the calculations are in a printout of an Excel preadsheet. How does this compare.